CIP-0177? | Ouroboros Tachys - Faster Cardano partner chains (#1149)
cardano-foundation/CIPs: CIP-0177/README.md

---
CIP: 177
Title: Ouroboros Tachys - Faster Cardano partner chains
Category: Consensus
Status: Proposed
Authors:
 - Duncan Coutts <duncan@well-typed.com>
 - Philipp Kant <philipp@ensurable.systems>
Implementors:
 - Well-Typed LLP
 - Ensurable Systems Ltd

> Discussions: https://github.com/cardano-foundation/CIPs/pull/1149

Reviewer: ```suggestion
Discussions:
 - Original PR: https://github.com/cardano-foundation/CIPs/pull/1149
```
(canonical YAML "bullet" list for GitHub rendering with explicit label)

The advantage of this Ouroboros variant would be a higher block production rate: approximately 4 times higher compared to Ouroboros Praos. This leads directly to 4x higher TPS and substantially lower transaction finality times. Although the protocol is different, for tools and applications that use the blockchain it would remain highly compatible, allowing the reuse of existing and future Cardano tooling.

The main disadvantage of Tachýs compared to Praos would be not having the DDoS resistance that Praos provides. For partner chains intended to have relatively few block producers, the DDoS resistance Praos provides is of limited benefit. Other forms of DDoS resistance are required, and are possible.

This CIP addresses the problems of settlement speed and transaction throughput for the use case of partner chains. These are related to the problems of settlement speed and transaction throughput for the Cardano mainnet identified in [CPS-0017] and [CPS-0018].

> ## Motivation: why is this CIP necessary?

Reviewer: ```suggestion
## Motivation: Why is this CIP necessary?
```
(fix long term capitalisation issue in template headings)

### Security Considerations

Security in blockchain systems emerges from the combined effect of protocol-level guarantees, engineering choices, and operational practices. No single layer is sufficient on its own. Tachýs makes a conscious set of trade-offs relative to Praos. In particular, it prioritises lower latency and higher throughput, and drops some protocol-level protections that arise from leader unpredictability. This section documents these trade-offs and their implications.

Implementers and operators would like clarity on which security properties the protocol guarantees and which it does not. With this clarity they can understand which risks must be addressed through engineering or operational measures, and which risks are already mitigated or eliminated at the protocol level. In this CIP we do not provide security proofs, however we set out the [scenarios and models](#proposed-security-models) in which we believe it is worth analysing the protocol.

#### Trade-offs relative to Praos

Compared to Praos, Tachýs has a few key trade-offs. We detail them in subsequent sub-sections, but in summary the trade-offs are as follows.

>  1. Tachýs is not secure against an [adaptive adversary](#adaptive-corruption), but we conjecture that it is secure against a non-adaptive adversary (as in Ouroboros Classic). We [argue below](#adaptive-corruption) that an adaptive adversary is an unrealistic assumption (in the operational setting of mainnet or a partner chain), and that assuming a non-adaptive adversary is more realistic.

Reviewer 2: The whole security considerations chapter is well written, thanks for adding it.

As you are stating here that "an adaptive adversary is an unrealistic assumption in the operational setting of [the Cardano] mainnet", does this imply we can use Tachýs for the Cardano mainnet without lowering it's practical security?

Author: Thanks.

I think generally we don't want to make big claims or proposals about mainnet, and certainly not in this CIP. I'm happy to trim mentions of mainnet if that helps to keep the focus.

That said, I think there's a good case for all of us (researchers, engineers, community) to re-analyse, revisit and rethink whether an adaptive adversary is a reasonable assumption. So far in my own thinking about it, I've not been able to think of any practical scenario where the adaptive adversary assumption corresponds to some real threat. Obviously absence of evidence is not evidence of absence, but it suggests it's worth looking at more closely. It's also not terribly surprising, given that what Cardano launched with -- which we all thought was perfectly good at the time -- was not secure against an adaptive adversary.

I think the thing we need to appreciate is that it's a trade-off: Praos was not a pure upgrade vs Classic. We gained resistance to an adaptive adversary, but with costs.

Tools that do independent validation of blocks, and specifically block headers would need to be updated before being used on a network in Tachýs mode. This would include alternative node implementations, and any data nodes that are fully-validating. Non-validating data nodes would be unaffected.

### Community consensus

[CIP-0001] calls for the presentation of evidence of consensus within the community and discussion of significant objections or concerns raised during the discussion.

At the stage of initial publication of this CIP, the proposed design has only been reviewed by a handful of experts (acknowledged below). We invite review and discussion of this proposed CIP on (or referenced from) the CIP PR itself, and we will endeavour to address reasonable objections, concerns and suggestions.

> ## Rationale: how does this CIP achieve its goals?

Reviewer: ```suggestion
## Rationale: How does this CIP achieve its goals?
```
(continued)

### Security Considerations

Security in blockchain systems emerges from the combined effect of protocol-level guarantees, engineering choices, and operational practices. No single layer is sufficient on its own. Tachýs makes a conscious set of trade-offs relative to Praos. In particular, it prioritises lower latency and higher throughput, and drops some protocol-level protections that arise from leader unpredictability. This section documents these trade-offs and their implications.

Implementers and operators would like clarity on which security properties the protocol guarantees and which it does not. With this clarity they can understand which risks must be addressed through engineering or operational measures, and which risks are already mitigated or eliminated at the protocol level. In this CIP we do not provide security proofs, however we set out the [scenarios and models](#proposed-security-models) in which we believe it is worth analysing the protocol.

#### Trade-offs relative to Praos

Compared to Praos, Tachýs has a few key trade-offs. We detail them in subsequent sub-sections, but in summary the trade-offs are as follows.

>  1. Tachýs is not secure against an [adaptive adversary](#adaptive-corruption), but we conjecture that it is secure against a non-adaptive adversary (as in Ouroboros Classic). We [argue below](#adaptive-corruption) that an adaptive adversary is an unrealistic assumption (in the operational setting of mainnet or a partner chain), and that assuming a non-adaptive adversary is more realistic.

Reviewer 3: Do you mean static instead of non-adaptive?

Author: I'm happy to use whichever is clearer, more appropriate. I did consider both terms, based on their use in the two Ouroboros papers. The Praos paper defines a static adversary as one that "corrupts a family of
parties at the outset of the protocol", whereas the Classic paper doesn't define such a term but it explains that it allows a limited form of dynamic corruption: either at the outset or after two epochs delay.

So my rationale for using "non-adaptive" is to try and better match the idea that it corresponds to the Classic meaning of limited dynamic corruption, as opposed to the very dynamic corruption in Praos. I didn't want to imply that it's only corruption at the outset of the protocol.

So I'm happy to take terminology advice.

>  2. Tachýs certainly has no more protection against grinding than Praos, and further analysis may show it has strictly less protection. We [conjecture below](#epoch-randomness-and-grinding-resistance) that, just as with Praos, this can be mitigated when Tachýs is deployed in combination with an effective anti-grinding mechanism, such as Phalanx.

Reviewer 3: @nhenin , is Phalanx still actively pursued?

Reviewer: @perturbing if the authors or other experts don't respond to this review facet before a potential merge, I don't think that would be a problem since "anti-grinding mechanism[s] such as Phalanx" are well documented in CIP-0161 & therefore a useful reference here even if work on Phalanx itself is/were discontinued.  (@nhenin @rrtoledo)

Reviewer 4: It’s actively being pursued indeed, and we were waiting for feedback that never came. Could we merge this PR? This increment was mainly cosmetic, with nothing major included.

Reviewer: Yes @nhenin I think we can now.  Editors (with @kantp observing at the last CIP meeting) had been waiting for @perturbing to sign off on it last week (since only @Ryun1 and I were at the meeting & this CIP is within Thomas' field of interest) but it's also true that all his feedback has been incorporated & his latest expression for the CIP was supportive.

Again in summary, a private leader schedule forces a DDoS attacker to target all (or almost all) nodes, but with a private leader schedule alone, such attacks could still prevent honest chain growth. On the other hand a combination of engineering and operational measures can also force a DDoS attacker to target all accessible nodes and in this case, even when all accessible nodes are suppressed, honest chain growth would continue. At best, such large scale attacks can temporarily prevent users from submitting transactions or receiving blocks but does not prevent consensus or honest chain growth. In conclusion, a private leader schedule is neither sufficient on its own, nor necessary to protect against attacks on liveness through DDoS attacks.

#### Proposed security models

We propose two scenarios, and corresponding assumptions for analysing Ouroboros Tachýs.

> Both sets of assumptions are based on Ouroboros Classic and thus have a non-adaptive adversary. We also include the Praos assumption that messages can be delayed by up to Delta slots (as this poses no difficulties).

Reviewer 3: ```suggestion
Both sets of assumptions are based on Ouroboros Classic and thus have a static adversary. We also include the Praos assumption that messages can be delayed by up to Delta slots (as this poses no difficulties).
```

Author: If we decide to switch terms then yeah I'll change it consistently throughout the security sub-section.
